### PR TITLE
Remove offline version and provide a KAR file instead

### PR DESCRIPTION
**Pull request by kaikreuzer**
This PR replaces the offline version by downloadable archives that contain all add-ons. So instead of having a separate distro, users can decide to download the add-ons once and put them in the addons folder, which immediately makes them available offline.

Additionally, all relevant artifacts that the online distro might need are now contained in a single Maven repo on Bintray. There is no need to have repo.eclipse.org, oss.jfrog.org, jcenter.org, etc. anymore (which especially means that there won't be any issues anymore if one of these repos is down as it regularly happens for repo.eclipse.org).

There will be another PR for openhab-core and openhab2-addons, which all have to be merged in sync, once ready.

Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@ThomDietrich Also for your information - this (again, sorry...) will have impact on the documentation, but as discussed in the past, having a single distro version should make it in general easier for the users.

**Comment by ThomDietrich** [comment]
(Getting up to speed) No problem. I love the change and will be happy to do needed changes to docs. Could you give a deadline?

**Comment by BClark09** [comment]
Based on what I understand from the files in this PR. I have made a branch that makes a start on the [appropriate changes](https://github.com/openhab/openhab-linuxpkg/compare/master...BClark09:remove-offline?expand=1&w=1) needed in the deb builds. The file paths are likely wrong, but I expect to find the appropriate files in the new mvn nightly version later.

**Comment by kaikreuzer** [comment]
> Could you give a deadline?

I hope to find the time to finish it over the weekend.

> but I expect to find the appropriate files in the new mvn nightly version later

No, the https://bintray.com/openhab/mvn/online-repo/2.0.0.nightly repo will really only host the bundles of the Karaf features for the distro to download (individually).
The distro (+ the new addons.kar), will be at the same place as today, i.e. releases at https://bintray.com/openhab/mvn/openhab-distro and snapshots at https://openhab.ci.cloudbees.com/job/openHAB-Distribution/lastSuccessfulBuild (or alternatively on https://oss.jfrog.org/oss-snapshot-local/org/openhab/distro/)

**Comment by kaikreuzer** [comment]
Ok, I was able to solve this problem, uff.
Now still have to do bug fixing and testing and many small improvements and clean-ups. Hope to have it all done tomorrow night.

@theoweiss & @BClark09 I assume that it is safe if I this gets merged (once finished) before the apt changes are done - merging this will only make the old apt repo disappear, but the new one will stay untouched, right? So once this is merged, you have a chance to also test the apt changes first, before updating the repo for everyone.

**Comment by theoweiss** [comment]
> Ok, I was able to solve this problem, uff.

Concrats!
I just removed the cloudbees build trigger (sorry the ui snapshot is in german):
linuxpkgbuildtriggerremoved

We'll have to add it again as soon as we've adopted the changes:
linuxpkgbuildtrigger

**Comment by BClark09** [comment]
[The branch I mentioned](https://github.com/openhab/openhab-linuxpkg/compare/master...BClark09:remove-offline?expand=1&w=1) earlier seems to be working and prepared for this change, all that is required is the correct download urls for each file.

Check out my [bintray repo](https://bintray.com/benclark/apt-repo2/openhab2/2.0.0~20170109095517) for the results. 

`sudo apt-get install openhab2` will remove openhab2-online or openhab2-offline automatically.
`sudo apt-get install openhab2-addons` or `sudo apt-get install openhab2-addons-legacy` will install openhab2 first.

**Comment by kaikreuzer** [comment]
Ok, I am done with my changes and would go ahead and merge them this afternoon.
We can then do final testing on the "standalone" distro before updating the apt package as well.

File: features/distro-kar/pom.xml [hunk 1]
```diff
@@ -32,20 +32,27 @@
...
-                    <startLevel>50</startLevel>
+                    <startLevel>80</startLevel>
                     <aggregateFeatures>true</aggregateFeatures>
                     <!--  <resolver>(obr)</resolver> -->
                     <checkDependencyChange>true</checkDependencyChange>
                     <failOnDependencyChange>false</failOnDependencyChange>
                     <logDependencyChanges>true</logDependencyChanges>
                     <overwriteChangedDependencies>true</overwriteChangedDependencies>
+                    <ignoreDependencyFlag>true</ignoreDependencyFlag>
+          			<archive>
```
Comment (maggu2810) [inline features/distro-kar/pom.xml]: You mixed spaces and tabs for indentation.

**Comment by kaikreuzer** [comment]
@theoweiss, @BClark09 & @ThomDietrich, here we go: https://openhab.ci.cloudbees.com/job/openHAB-Distribution/

Please help with some testing and start preparing the next steps wrt apt & docs!

**Comment by BClark09** [comment]
Seems to be working well on install, legacy files are immediately available on placing the legacy kar file in the addons folder. Removing them crashes openhab however and the UI is stuck on this page until restarting:

image

```
2017-01-09 15:19:32.128 [WARN ] [org.apache.felix.fileinstall        ] - Failed to uninstall artifact: /usr/share/openhab2/addons/openhab-addons-2.0.0-SNAPSHOT.kar
java.lang.IllegalArgumentException: Feature named 'openhab-binding-hue/2.0.0.SNAPSHOT' is not installed
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.uninstallFeatures(FeaturesServiceImpl.java:997)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.uninstallFeatures(FeaturesServiceImpl.java:872)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.removeRepository(FeaturesServiceImpl.java:528)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.removeRepository(FeaturesServiceImpl.java:491)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.kar.internal.KarServiceImpl.uninstall(KarServiceImpl.java:232)[49:org.apache.karaf.kar.core:4.0.8]
        at org.apache.karaf.deployer.kar.KarArtifactInstaller.uninstall(KarArtifactInstaller.java:56)[41:org.apache.karaf.deployer.kar:4.0.8]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.uninstall(DirectoryWatcher.java:1070)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.uninstall(DirectoryWatcher.java:888)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:483)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:361)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:312)[4:org.apache.felix.fileinstall:3.5.6]
2017-01-09 15:19:33.136 [WARN ] [org.apache.felix.fileinstall        ] - Failed to uninstall artifact: /usr/share/openhab2/addons/openhab-addons-legacy-2.0.0-SNAPSHOT.kar
java.lang.IllegalArgumentException: Feature named 'openhab-binding-homematic1/1.9.0.SNAPSHOT' is not installed
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.uninstallFeatures(FeaturesServiceImpl.java:997)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.uninstallFeatures(FeaturesServiceImpl.java:872)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.removeRepository(FeaturesServiceImpl.java:528)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.removeRepository(FeaturesServiceImpl.java:491)[8:org.apache.karaf.features.core:4.0.8]
        at org.apache.karaf.kar.internal.KarServiceImpl.uninstall(KarServiceImpl.java:232)[49:org.apache.karaf.kar.core:4.0.8]
        at org.apache.karaf.deployer.kar.KarArtifactInstaller.uninstall(KarArtifactInstaller.java:56)[41:org.apache.karaf.deployer.kar:4.0.8]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.uninstall(DirectoryWatcher.java:1070)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.uninstall(DirectoryWatcher.java:888)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.doProcess(DirectoryWatcher.java:483)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.process(DirectoryWatcher.java:361)[4:org.apache.felix.fileinstall:3.5.6]
        at org.apache.felix.fileinstall.internal.DirectoryWatcher.run(DirectoryWatcher.java:312)[4:org.apache.felix.fileinstall:3.5.6]
```
Neither Hue, nor Homematic were ever installed in my test, so not sure why it picks them out.

**Comment by hakan42** [comment]
So how does the installation work now? Do I put the complete kar file(s) into the addons folder?

**Comment by BClark09** [comment]
Yes, exactly that. For apt-get users there will be a package that does this and updates for you.

**Comment by kaikreuzer** [comment]
Removal is no supported usecase :-)
This is a bug in the Karaf KAR deployer, that I also noticed already. Needs to be addressed in Karaf, I will try to report a bug, as soon as I find time.

**Comment by BClark09** [comment]
Alrighty then! Users should be fine in restarting openHAB in the rare case that they do that ;). The new openhab2, openhab2-addons and openhab2-addons-legacy packages worked fine, so have prepared https://github.com/openhab/openhab-linuxpkg/pull/18 for review and testing.

**Comment by theoweiss** [comment]
I've merged @BClark09's PR and reconfigured cloudbees to trigger the linuxpkg build on successful openHAB-Distribution builds. Currently an openHAB-Distribution build is running, we'll see if the liunxpkg also works.

**Comment by kaikreuzer** [comment]
> Currently an openHAB-Distribution build is running

But it does not seem to want to finish... It is already running since an hour, doing the upload to Bintray :-/
I'd suggest you simply trigger your build manually :-)

**Comment by theoweiss** [comment]
Done. The packages are at Bintray. We'll have to check it tomorrow.

**Comment by BClark09** [comment]
I can check now, I'll do that and the post on the community page if that's okay?

**Comment by ThomDietrich** [comment]
I've just upgraded from old repository to new plus switching over from `openhab2-offline` to `openhab2`. Everything worked like a charm. Thanks guys!
I've updated openHABian to install `openhab2` from here on out.

**Pull request merged**